### ISSUE TITLE
Nextflow?

### ISSUE THREAD

**Issue by SteffenBrinckmann** [NFDI4Chem/formaTAPIRest#6]
Hey,
How about we define each conversion as a nextflow (www.nextflow.io) workflow: take input, do container work, produce output.
- Then all conversions become a list of workflows, incl. their metadata (instrument, vendor), which can be easily searched.
- TAPIR then can be an API, that wraps (the list of workflows and the nextflow system)
- The advantage, we can use nextflow which is widely distributed and don't have to create a system for converter definition to ourselves. There are people that can help.
- Vendors can follow the system and create their own workflows

Our task would be to define example converters:
- we have already png->jpg and mov->mp4 would allow to test large data
- output to file and output stream might be favorable in different cases: if there is lots of data, (mov->mp4) writing first to file is slow.

What do you think?
Steffen

**Comment by mktudr** [comment]
Hi,
Steffen N. and me discussed very briefly existing workflows like nextflow and galaxyproject (https://galaxyproject.org/learn/advanced-workflow/). At the moment we didn't want to be dependent on a large project in order to remain flexible, although we also need some of the functions. But Steffen N. mentioned current work on a common workflow language that we might use. I'm not sure, but maybe this is interesting: https://en.wikipedia.org/wiki/Common_Workflow_Language.

**Comment by sneumann** [comment]
Indeed, CWL has quite a momentum also within Galaxy and Nextflow communities. I looked at the command-line reference implementation https://github.com/common-workflow-language/cwltool#run-on-the-command-line to keep dependencies slim. Currently I am lacking the big picture, e.g. whether they have a common toolsched for the tool definitions. Indeed, one could say we're currently re-designing something like https://github.com/SciDAS/nextflow-api . Yours, Steffen

**Comment by holgerisrael** [comment]
Does any of you know how interoperable Nextflow and CWR are? 

I've had a very brief look at the Nextflow documentation, and what I get is that what it does it to talk to the batch scheduling codes, and also the containerisation softwares, if I get it right...

By the way: In the NFDI4Phys Extended Abstract, we're referencing the "Canonical Workflow Framework for Research" initiative, which looks like it's an abstraction of CWR... 
https://codata.org/wp-content/uploads/2021/01/CWFR-position-paper-v3.pdf

**Comment by SteffenBrinckmann** [comment]
I kind of like the workflow idea as it uses a defined system for conversion. I also like the CWL, as it makes us "platform independent". The question then would be, which features does the platform require:
- REST API
And the CWL can store the meta-data (Instrument description, Author, ...) and link to the conversion (.py, docker...)
Would you like to use CWL? Do we want to think about it for a week or so?

**Comment by SteffenBrinckmann** [comment]
The other BIG advantage of CWL: testing
- the person that writes the converter can test it before submitting
- the system can test it during submission with an example file (which is also submitted) and then both are stored.
(I am writing some examples right now and have little idea if they work because I might have forgotten a python library)

**Comment by holgerisrael** [comment]
The ability to test indeed sounds like a BIG advantage to me... (Recalling my experience as a user of a pipeline whose orchestration nobody understood any longer :-( )
So I'm curious about testing CWL - but I guess the question was more for Steffen N and Michael...

**Comment by sneumann** [comment]
So if we follow a strategy of incremental changes, I'd love someone ( :-) ) to take a first shot with the png2jpg, and I'll adopt the msconvert example from there. msconvert already has a [Galaxy tool](https://toolshed.g2.bx.psu.edu/repository?repository_id=d2409f4a98066d94&changeset_revision=001e77708ec9) and a [NextFLow component](https://github.com/lehtiolab/nf-msconvert).

**Comment by SteffenBrinckmann** [comment]
Would it be good to first decide on CWL or NextFlow? Galaxy seems to have CWL integration, NextFlow not. Nextflow seems more fancy but CWL more standard.

**Comment by SteffenBrinckmann** [comment]
I read some of the CWL docs and will use Monday to work towards some examples. I will add them in a separate folder with its own Readme.md. If that is ok.

**Comment by SteffenBrinckmann** [comment]
Hey,
Do we want to meet and talk about the next steps?
-which meta data do we want?
-is a github folder with all cwl files (plus CI to create overview table
enough)?
- background: one can post recipe.cwl and input file to server (eg arvados)
and get result back.
What do you think?
Best, Steffen

On Jun 10, 2021 at 13:56, Steffen Neumann ***@***.***> wrote:

So if we follow a strategy of incremental changes, I'd love someone ( :-) )
to take a first shot with the png2jpg, and I'll adopt the msconvert example
from there. msconvert already has a Galaxy tool
<https://toolshed.g2.bx.psu.edu/repository?repository_id=d2409f4a98066d94&changeset_revision=001e77708ec9>
and a NextFLow component <https://github.com/lehtiolab/nf-msconvert>.

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub
<https://github.com/NFDI4Chem/formaTAPIRest/issues/6#issuecomment-858557411>,
or unsubscribe

**Comment by sneumann** [comment]
So some more pointers to related projects:
- https://github.com/MD-Studio/cerise is a REST service for running CWL workflows on remote clusters => The generic CWL version of what we're working on here. Without OpenAPI UI. 
- https://github.com/Duke-GCB/calrissian => running CWL on a kubernetes cluster, so basically the replacement for my makeshift Kubernetes Adapter stolen from Galaxy.

**Comment by sneumann** [comment]
I still need to wrap my mind around the coolest way to integrate CWL in our architecture:
- The formaTAPIRest project includes a directory `workflows/` containing `*.cwl` tool wrappers, e.g. `msconvert.cwl`
- Instead of assembling the command line by hand, the formaTAPIRest would create the `input.yml` from REST input, and call `cwl-runner`. 

With `cwl-runner` we can keep 1) local execution on the service host and 2) local docker , we gain other container systems like singularity. For Kubernetes I need to figure out Calrissian.
